### PR TITLE
disable cfg mode  when binary filter is set

### PR DESCRIPTION
**Pull request by cppchedy**
disabling cfg mode when binary filter is set.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,7 +16,7 @@
   },
```
Comment (jaredwy) [inline package.json]: Could you revert the changes in this file.

File: static/cfg-view.js [hunk 2]
```diff
@@ -156,6 +157,16 @@ define(function (require) {
             this.saveState();
         }
     };
+    
+    Cfg.prototype.closeCfg = function () {
+         _.defer(function (self) {
```
Comment (jaredwy) [inline static/cfg-view.js]: Why the defer?

This can be rewritten to be _.defer(this.container.close);

File: static/cfg-view.js [hunk 3]
```diff
@@ -156,6 +157,16 @@ define(function (require) {
...
+                self.container.close();
+            }, this);
+    };
+    
+    Cfg.prototype.onBinaryMode = function() {
```
Comment (jaredwy) [inline static/cfg-view.js]: Why not just call closeCfg in the binding instead of having this indirection?

**Comment by jaredwy** [comment]
Instead of closing we have been putting in "this is not supported" if it is open. 
Try opening the opt-view then changing your compiler to gcc. 
Could you do that instead?

File: static/cfg-view.js [hunk 4]
```diff
@@ -156,6 +157,16 @@ define(function (require) {
...
+        this.closeCfg();
```
Comment (RabsRincon) [inline static/cfg-view.js]: I don't think closing the cfg is the way to go here. Maybe update the graph to only say "CFG not supported on binary mode" liek we do when no output is found

File: package.json [hunk 5]
```diff
@@ -16,7 +16,7 @@
...
   "main": "./app.js",
   "dependencies": {
-    "aws-sdk": "*",
+    "aws-sdk": "^2.122.0",
```
Comment (RabsRincon) [inline package.json]: Is this an upstream merge I can't find or did you just edit them for a reason?

File: static/compiler.js [hunk 6]
```diff
@@ -678,7 +678,10 @@ define(function (require) {
 
     Compiler.prototype.onCfgViewClosed = function (id, network) {
         if (this.id == id) {
-            this.cfgButton.prop('disabled', false);
+            var state = false;
+            if (this.filters.binary) 
+                state = true;
+            this.cfgButton.prop('disabled', state);
```
Comment (RabsRincon) [inline static/compiler.js]: This can all be simplified to `this.cfgButton.prop('disabled', this.filters.binary);`

File: package.json (hunk 1 above)
Comment (cppchedy) [reply package.json]: ok

File: static/compiler.js [hunk 7]
```diff
@@ -752,6 +755,9 @@ define(function (require) {
         this.saveState();
         this.compile();
         this.updateButtons();
+        if(this.getEffectiveFilters().binary){
+            this.eventHub.emit('binaryMode');
```
Comment (RabsRincon) [inline static/compiler.js]: Pass the compiler id for which the binaryMode was toggled and its new state so not every cfg from every compiler gets updated when only one does

File: static/cfg-view.js (hunk 2 above)
Comment (cppchedy) [reply static/cfg-view.js]: ok

File: static/cfg-view.js (hunk 3 above)
Comment (cppchedy) [reply static/cfg-view.js]: following naming convention onEventName but this naming doesn't tell what's the function is doing. so I added one level.

File: static/cfg-view.js (hunk 4 above)
Comment (cppchedy) [reply static/cfg-view.js]: right! it's a better idea!

File: static/compiler.js (hunk 6 above)
Comment (cppchedy) [reply static/compiler.js]: pfffff! I hate it when I don't see obvious things like that!

File: static/cfg-view.js [hunk 8]
```diff
@@ -37,7 +37,8 @@ define(function (require) {
         this.eventHub = hub.createEventHub();
         this.domRoot = container.getElement();
         this.domRoot.html($('#cfg').html());
-        this.defaultCfgOutput = {nodes: [{id: 0, label: 'No Output'}], edges: []};
+        this.defaultCfgOutput = {nodes: [{id: 0, shape:"box", label: 'No Output'}], edges: []};
+        this.binaryModeSupport = {nodes: [{id: 0, shape:"box", label: "Cfg mode cannot be used when binary filter is set"}], edges: []};
```
Comment (RabsRincon) [inline static/cfg-view.js]: Minor: 
```diff
- cannot be used when binary filter is set
+ cannot be used when the binary filter is set
```

File: static/cfg-view.js [hunk 9]
```diff
@@ -127,7 +128,7 @@ define(function (require) {
         this.setTitle();
     }
 
-    Cfg.prototype.onCompileResult = function (id, compiler, result) {
+    Cfg.prototype.onCompileResult = function (id, compiler, result, binaryFilter) {
```
Comment (RabsRincon) [inline static/cfg-view.js]: As discussed on slack, a "filtersChange" which passes a compilerId and the effective filters is probably a better idea than modifying onCompileResult
Also, Jared's suggestion of only sending the diff would probably be a good idea but I think it falls out of scope for the PR

File: static/compiler.js [hunk 10]
```diff
@@ -620,7 +620,7 @@ define(function (require) {
             compileTime.text("");
         }
         this.compilerSupportsCfg = result.supportsCfg;
-        this.eventHub.emit('compileResult', this.id, this.compiler, result);
+        this.eventHub.emit('compileResult', this.id, this.compiler, result, this.getEffectiveFilters().binary);
```
Comment (RabsRincon) [inline static/compiler.js]: As above, modifying this is prob not the way to go

File: static/compiler.js [hunk 11]
```diff
@@ -676,9 +676,9 @@ define(function (require) {
         }
     };
 
-    Compiler.prototype.onCfgViewClosed = function (id, network) {
+    Compiler.prototype.onCfgViewClosed = function (id) {
```
Comment (RabsRincon) [inline static/compiler.js]: Nice spot! I forgot to remove the arg when I move the destroy funcitonality back to the view itself

File: static/cfg-view.js [hunk 12]
```diff
@@ -187,7 +205,8 @@ define(function (require) {
             id: this._compilerid,
             editorid: this._editorid,
             selectedFn: this.currentFunc,
-            compilerName: this._compilerName
+            compilerName: this._compilerName,
+            binaryFilter: this._binaryFilter
```
Comment (RabsRincon) [inline static/cfg-view.js]: Why do you need to save the binary filters here?

File: static/cfg-view.js [hunk 13]
```diff
@@ -171,6 +174,21 @@ define(function (require) {
             this.setTitle();
         }
     };
+    
+    Cfg.prototype.onFiltersChange = function (id, filters) {
+        if (this._compilerid === id) {
+            if (filters.binary !== this._binaryFilter) {
+                if (filters.binary === true) {
+                    new Alert().notify("Cfg mode cannot be used when the binary filter is set.", {
```
Comment (RabsRincon) [inline static/cfg-view.js]: Is an alert a better alternative to displaying this on the cfg view itself?

File: static/compiler.js [hunk 14]
```diff
@@ -275,7 +275,8 @@ define(function (require) {
         }, this);
 
         var createCfgView = _.bind(function () {
-            return Components.getCfgViewWith(this.id, this.source, this.lastResult.cfg, this.getCompilerName(), this.sourceEditorId);
+            return Components.getCfgViewWith(this.id, this.source, this.lastResult.cfg, this.getCompilerName(), this.sourceEditorId, 
+                                             this.getEffectiveFilters().binary);
```
Comment (RabsRincon) [inline static/compiler.js]: Again, is this a better approach than asking for the filters once the cfg view has been innited?

File: static/cfg-view.js [hunk 15]
```diff
@@ -171,6 +175,14 @@ define(function (require) {
             this.setTitle();
         }
     };
+    
+    Cfg.prototype.onFiltersChange = function (id, filters) {
+        if (this._compilerid === id) {
+            if (filters.binary !== this._binaryFilter) {
```
Comment (jaredwy) [inline static/cfg-view.js]: Why is this needed?

Comment (cppchedy) [reply static/cfg-view.js]: not needed.
